src/concrete/ml/sklearn/tree_to_numpy.py

```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+
+    def get_lsbs_to_remove(array: numpy.ndarray) -> int:
+        """Update the number of LSBs to remove based on overflow detection.
+
+        Args:
+            array (umpy.ndarray): The array for which the bitwidth needs to be checked.
+
+        Returns:
+            int: The updated LSB to remove.
+        """
+
+        initial_bitwidth = get_bitwidth(array)
+        # The subtraction operation increases precision by 1 or 2 bits
+        lsbs_to_remove = initial_bitwidth + 2
+
+        while lsbs_to_remove > 0:
+            half = 1 << (lsbs_to_remove - 1)
+            if get_bitwidth(array - half) <= initial_bitwidth:
+                lsbs_to_remove -= 1  # pragma: no cover
+            else:
+                break  # pragma: no cover
+
+        return lsbs_to_remove
+
+    quant_params = {
+        onnx_init.name: numpy_helper.to_array(onnx_init)
+        for onnx_init in onnx_model.graph.initializer
+        if "weight" in onnx_init.name or "bias" in onnx_init.name
+    }
+
+    key_mat_1 = [key for key in quant_params.keys() if "_1" in key and "weight" in key][0]
+    key_bias_1 = [key for key in quant_params.keys() if "_1" in key and "bias" in key][0]
+
+    key_mat_2 = [key for key in quant_params.keys() if "_2" in key and "weight" in key][0]
+    key_bias_2 = [key for key in quant_params.keys() if "_2" in key and "bias" in key][0]
+
+    # shape: (nodes, features) or (trees * nodes, features)
+    mat_1 = quant_params[key_mat_1]
+    # shape: (nodes, 1) or (trees * nodes, 1)
+    bias_1 = quant_params[key_bias_1]
+
+    # shape: (trees, leaves, nodes)
+    mat_2 = quant_params[key_mat_2]
+    # shape: (leaves, 1) or (trees * leaves, 1)
+    bias_2 = quant_params[key_bias_2]
+
+    n_features = mat_1.shape[1]
+    n_nodes = mat_2.shape[2]
+    n_leaves = mat_2.shape[1]
+
+    mat_1 = mat_1.reshape(-1, n_nodes, n_features)
+    bias_1 = bias_1.reshape(-1, 1, n_nodes)
+    bias_2 = bias_2.reshape(-1, 1, n_leaves)
+
+    required_onnx_operators = set(get_op_type(node) for node in onnx_model.graph.node)
+
+    # If operator is `<`, np.less(x, y) is equivalent to:
+    # round_bit_pattern((x - y) - half, lsbs_to_remove=r) < 0.
+    # Therfore, stage_1 = (q_x @ mat_1.transpose(0, 2, 1)) - bias_1
+    if "Less" in required_onnx_operators:
+        stage_1 = (q_x @ mat_1.transpose(0, 2, 1)) - bias_1
+        matrix_q = stage_1 < 0
+    elif "LessOrEqual" in required_onnx_operators:
+        # If operator is `<=`, np.less_equal(x, y) is equivalent to:
+        # round_bit_pattern((y - x) - half, lsbs_to_remove=r) >= 0.
+        # Therfore, stage_1 = bias_1 - (q_x @ mat_1.transpose(0, 2, 1))
+        stage_1 = bias_1 - (q_x @ mat_1.transpose(0, 2, 1))
+        matrix_q = stage_1 >= 0
+
+    print("sum ****************** ", matrix_q.shape, matrix_q.sum())
+
+    lsbs_to_remove_stage_1 = get_lsbs_to_remove(stage_1)
+
+    # If operator is `==`, np.equal(x, y) is equivalent to:
+    # round_bit_pattern((x - y) - half, lsbs_to_remove=r) >= 0.
```

Comment:
still not sure why we have this `>=0` instead of `==0`

Reply:
because, when x is close to y with a tolerance of `half` x  - y - half  == 0 is not equivalent to 
```
        unit = 1 << lsbs_to_remove
        half = 1 << lsbs_to_remove - 1
        rounded = (x + half) // unit
        return rounded * unit
```